net.c

```diff
@@ -7,143 +7,237 @@
-        twarn("setting SO_KEEPALIVE on fd %d", fd);
-        close(fd);
-        continue;
-      }
-      r = setsockopt(fd, SOL_SOCKET, SO_LINGER, &linger, sizeof linger);
-      if (r == -1) {
-        twarn("setting SO_LINGER on fd %d", fd);
-        close(fd);
-        continue;
-      }
-      r = setsockopt(fd, IPPROTO_TCP, TCP_NODELAY, &flags, sizeof flags);
-      if (r == -1) {
-        twarn("setting TCP_NODELAY on fd %d", fd);
+        r = bind(fd, ai->ai_addr, ai->ai_addrlen);
+        if (r == -1) {
+            twarn("bind()");
+            close(fd);
+            continue;
+        }
+        if (verbose) {
+            char hbuf[NI_MAXHOST], pbuf[NI_MAXSERV], *h = host, *p = port;
+            struct sockaddr_in addr;
+            socklen_t addrlen;
+
+            addrlen = sizeof(addr);
+            r = getsockname(fd, (struct sockaddr *) &addr, &addrlen);
+            if (!r) {
+                r = getnameinfo(ai->ai_addr, ai->ai_addrlen,
+                                hbuf, sizeof hbuf,
+                                pbuf, sizeof pbuf,
+                                NI_NUMERICHOST|NI_NUMERICSERV);
+                if (!r) {
+                    h = hbuf;
+                    p = pbuf;
+                }
+            }
+            if (ai->ai_family == AF_INET6) {
+                printf("bind %d [%s]:%s\n", fd, h, p);
+            } else {
+                printf("bind %d %s:%s\n", fd, h, p);
+            }
+        }
+
+        r = listen(fd, 1024);
+        if (r == -1) {
+            twarn("listen()");
+            close(fd);
+            continue;
+        }
+
+        break;
+    }
+
+    freeaddrinfo(airoot);
+
+    if(ai == NULL)
+        fd = -1;
+
+    return fd;
+}
+
+static int
+make_unix_socket(char *path)
+{
+    int fd = -1, r;
+    struct stat st;
+    struct sockaddr_un addr;
+    const size_t maxlen = sizeof(addr.sun_path) - 1; // Reserve the last position for '\0'
+
+    r = stat(path, &st);
+    if (r == 0) {
+        if (S_ISSOCK(st.st_mode)) {
+            warnx("removing existing local socket to replace it");
+            r = unlink(path);
+            if (r == -1) {
+                twarn("unlink");
+                return -1;
+            }
+        } else {
+            twarnx("another file already exists in the given path");
+            return -1;
+        }
+    }
+
+    fd = socket(AF_UNIX, SOCK_STREAM, 0);
+    if (fd == -1) {
+        twarn("socket()");
+        return -1;
+    }
+
+    r = set_nonblocking(fd);
+    if (r == -1) {
         close(fd);
-        continue;
-      }
+        return -1;
+    }
 
-      r = bind(fd, ai->ai_addr, ai->ai_addrlen);
-      if (r == -1) {
+    memset(&addr, 0, sizeof(struct sockaddr_un));
+    addr.sun_family = AF_UNIX;
+    if (strlen(path) > maxlen) {
+        warnx("truncating local socket path to: %.*s", (int)maxlen, path);
+    }
+    strncpy(addr.sun_path, path, maxlen);
+
+    r = bind(fd, (struct sockaddr *) &addr, sizeof(struct sockaddr_un));
```

Comment:
I guess you meant `struct sockaddr_un` in both of casts.

Reply:
No, the cast is intentional (otherwise it wouldn't be necessary), see [bind(2)'s man page][1].

- Simple answer: it wouldn't compile otherwise, because warnings are treated as errors
- Longer answer: this is C-style poor man's polymorphism from the 80's, and also why the structure length needs to be provided too

[1]: https://manpage.me/index.cgi?q=bind&sektion=2